Data/ByteString/Char8.hs

```diff
@@ -623,6 +625,24 @@ index :: ByteString -> Int -> Char
 index = (w2c .) . B.index
 {-# INLINE index #-}
 
+-- | /O(1)/ 'ByteString' index, starting from 0, that returns 'Just' if:
+--
+-- > 0 <= n < length bs
+--
+-- @since 0.10.10.1
+indexMaybe :: ByteString -> Int -> Maybe Char
+indexMaybe = ((fmap w2c) .) . B.indexMaybe
```

Comment:
Out of curiosity:

Is there a (performance?) reason for preferring this IMHO rather unreadable style instead of something like

```suggestion
indexMaybe bs ix = fmap w2c (B.indexMaybe bs ix)
```

Something related to inlining perhaps?

Reply:
Indeed this  η-reduced form is expected to inline under more circumstances, but IIRC the same can be achieved via λ-abstraction, perhaps to an intermediate form, such as:
```haskell
indexMaybe = \ bs -> fmap w2c . B.indexMaybe bs
```
which for me is still quite readable (the omitted `ix` is easily inferred), but dropping `bs` has a much greater impact on readability.  Of course one might then simply write
```haskell
indexMaybe = \ bs ix -> fmap w2c $ B.indexMaybe bs ix
```
FWIW, one might also write the η-reduced form as:
```haskell
indexMaybe = (fmap w2c .) . B.indexMaybe
```
thereby avoiding one redundant pair of parentheses.

Reply:
Errr, because of how ghc optimizes, it’s definitely possible to arrange stuff so that code written in point free style will fail to inline when the equivalent lambda / applied / $ form will. Not due to anything deep, but simply due to the local way ghc decided when a function is fully applied “saturated/ saturation condition”.  

 Or maybe I’m missing the point?

Ohhhh

This is about writing f x =...
 vs 
f = \ x -> ... 

For a definition! 

I favor explicit lambda more than point free style for that.  I actually developed it into a whole style in my numerical repo a few years back

Reply:
Well, the original review comment was about `f x = ` vs. point-free style...  That said, I was under the impression that for inlining, point-free `f = g` is equivalent to `f = \ x -> g x`, and that only `f x = g x` might fail to inline at call sites using just `f` with no explicit arguments (e.g. `h = f . g`).  Was I wrong about that?

Reply:
It's better than `indexMaybe xs i = fmap w2c $ B.indexMaybe xs i` at least, but I think `indexMaybe xs = fmap w2c . B.indexMaybe xs` is the sweet spot.